### ISSUE TITLE
Add ADAM

### ISSUE THREAD

**Issue by leonardoaraujosantos** [leonardoaraujosantos/DLMatFramework#13]
## Introduction
Add to the Solver class a member that can do Adam (Inherited from Optimiser class).

## Point of improvement
Some optimisers like SGD with momentum or Adam have a particular state (velocity on SGD with momentum) for each layer. On issue #12 I've just create those members (One for bias and the other for weights) directly on the Solver class, which is lame. So improve it now.

## Algorithm description
![screen shot 2017-03-02 at 12 19 40](https://cloud.githubusercontent.com/assets/898383/23506943/a551c32a-ff42-11e6-8be1-69bcd093607a.png)

## Comparison with other methods
![adam2](https://cloud.githubusercontent.com/assets/898383/23311741/11e7c2e8-faaf-11e6-855a-a30830c113bc.png)

## Code reference (cs231n)
``` python
def adam(x, dx, config=None):
  """
  Uses the Adam update rule, which incorporates moving averages of both the
  gradient and its square and a bias correction term.
  config format:
  - learning_rate: Scalar learning rate.
  - beta1: Decay rate for moving average of first moment of gradient.
  - beta2: Decay rate for moving average of second moment of gradient.
  - epsilon: Small scalar used for smoothing to avoid dividing by zero.
  - m: Moving average of gradient.
  - v: Moving average of squared gradient.
  - t: Iteration number.
  """
  if config is None: config = {}
  config.setdefault('learning_rate', 1e-3)
  config.setdefault('beta1', 0.9)
  config.setdefault('beta2', 0.999)
  config.setdefault('epsilon', 1e-8)
  config.setdefault('m', np.zeros_like(x))
  config.setdefault('v', np.zeros_like(x))
  config.setdefault('t', 0)
  
  next_x = None

  # Update here...
  config['t']+=1
  config['m'] = config['beta1']*config['m'] + (1- config['beta1'])*dx
  config['v'] = config['beta2']*config['v'] + (1- config['beta2'])*(dx**2)
  mb = config['m']/(1-config['beta1']**config['t'])
  vb = config['v']/(1-config['beta2']**config['t'])
  next_x = x -config['learning_rate']* mb / (np.sqrt(vb) + config['epsilon'])
  
  return next_x, config
```

## References
* https://arxiv.org/pdf/1412.6980.pdf
* http://cs231n.github.io/neural-networks-3/#ada
* http://sebastianruder.com/optimizing-gradient-descent/index.html#adam
* http://cs229.stanford.edu/proj2015/054_report.pdf
* https://moodle2.cs.huji.ac.il/nu15/pluginfile.php/316969/mod_resource/content/1/adam_pres.pdf
* https://www.quora.com/Can-you-explain-basic-intuition-behind-ADAM-a-method-for-stochastic-optimization

**Comment by leonardoaraujosantos** [comment]
## Some Results
Some tests with FC network solving the MNIST dataset
``` matlab
clear all;
rng(0,'v5uniform');
%% Load data
load mnist_oficial;
data = Dataset(input_train, output_train_labels,1,784,1,1,true);
data.AddValidation(input_test,output_test_labels,1,784,1,1,true);

%% Create network
layers = LayerContainer();    
layers <= struct('name','ImageIn','type','input','rows',1,'cols',784,'depth',1, 'batchsize',1);
layers <= struct('name','FC_1','type','fc', 'num_output',200);
layers <= struct('name','BN_1','type','batchnorm','eps',1e-5, 'momentum', 0.9);
layers <= struct('name','Relu_1','type','relu');
layers <= struct('name','FC_2','type','fc', 'num_output',100);
layers <= struct('name','BN_2','type','batchnorm','eps',1e-5, 'momentum', 0.9);
layers <= struct('name','Relu_2','type','relu');
layers <= struct('name','FC_3','type','fc', 'num_output',50);
layers <= struct('name','BN_3','type','batchnorm','eps',1e-5, 'momentum', 0.9);
layers <= struct('name','Relu_3','type','relu');
layers <= struct('name','DRP_1','type','dropout','prob',0.5);
layers <= struct('name','FC_4','type','fc','num_output',data.GetNumClasses());
layers <= struct('name','Softmax','type','softmax');

% Create DeepLearningModel instance
net = DeepLearningModel(layers, LossFactory.GetLoss('multi_class_cross_entropy'));


%% Create solver and train
solver = Solver(net, data, 'adam',containers.Map({'learning_rate', 'L2_reg'}, {0.01, 0}));
solver.SetBatchSize(200);
solver.SetEpochs(60);
solver.Train();
```

### MNIST FC
#### SGD Vanilla 
![sgd_vanilla](https://cloud.githubusercontent.com/assets/898383/23310486/9e9524b0-faaa-11e6-8f08-bfbd96fb9165.png)

#### With Momentum
![sgd_momentum](https://cloud.githubusercontent.com/assets/898383/23310264/e29a2f6c-faa9-11e6-9a29-214eb3a2b96e.png)

#### With Adam
![adam_fc_mnist](https://cloud.githubusercontent.com/assets/898383/23611454/63277b5e-026f-11e7-8afd-b400016edaa5.png)

**Comment by leonardoaraujosantos** [comment]
Seems ok for now...
